x-pack/packages/ml/agg_utils/src/fetch_histograms_for_fields.ts

```diff
@@ -0,0 +1,233 @@
+
+import { KBN_FIELD_TYPES } from '@kbn/field-types';
+import { isPopulatedObject } from '@kbn/ml-is-populated-object';
+import { stringHash } from '@kbn/ml-string-hash';
+
+import { buildSamplerAggregation } from './build_sampler_aggregation';
+import { fetchAggIntervals } from './fetch_agg_intervals';
+import { getSamplerAggregationsResponsePath } from './get_sampler_aggregations_response_path';
+import type {
+  AggCardinality,
+  ElasticsearchClient,
+  HistogramField,
+  NumericColumnStats,
+  NumericColumnStatsMap,
+} from './types';
+
+const MAX_CHART_COLUMNS = 20;
+
+interface AggHistogram {
+  histogram: {
+    field: string;
+    interval: number;
+  };
+}
+
+interface AggTerms {
+  terms: {
+    field: string;
+    size: number;
+  };
+}
+
+interface NumericDataItem {
+  key: number;
+  key_as_string?: string;
+  doc_count: number;
+}
+
+export interface NumericChartData {
+  data: NumericDataItem[];
+  id: string;
+  interval: number;
+  stats: [number, number];
+  type: 'numeric';
+}
+
+export interface NumericHistogramField extends HistogramField {
+  type: KBN_FIELD_TYPES.DATE | KBN_FIELD_TYPES.NUMBER;
+}
+type NumericHistogramFieldWithColumnStats = NumericHistogramField & NumericColumnStats;
+
+function isNumericHistogramField(arg: unknown): arg is NumericHistogramField {
+  return (
+    isPopulatedObject(arg, ['fieldName', 'type']) &&
+    (arg.type === KBN_FIELD_TYPES.DATE || arg.type === KBN_FIELD_TYPES.NUMBER)
+  );
+}
+function isNumericHistogramFieldWithColumnStats(
+  arg: unknown
+): arg is NumericHistogramFieldWithColumnStats {
+  return (
+    isPopulatedObject(arg, ['fieldName', 'type', 'min', 'max', 'interval']) &&
+    (arg.type === KBN_FIELD_TYPES.DATE || arg.type === KBN_FIELD_TYPES.NUMBER)
+  );
+}
+
+interface OrdinalDataItem {
+  key: string;
+  key_as_string?: string;
+  doc_count: number;
+}
+
+interface OrdinalChartData {
+  type: 'ordinal' | 'boolean';
+  cardinality: number;
+  data: OrdinalDataItem[];
+  id: string;
+}
+
+interface OrdinalHistogramField extends HistogramField {
+  type: KBN_FIELD_TYPES.STRING | KBN_FIELD_TYPES.BOOLEAN;
+}
+
+function isOrdinalHistogramField(arg: unknown): arg is OrdinalHistogramField {
+  return (
+    isPopulatedObject(arg, ['fieldName', 'type']) &&
+    (arg.type === KBN_FIELD_TYPES.STRING || arg.type === KBN_FIELD_TYPES.BOOLEAN)
+  );
+}
+
+interface UnsupportedChartData {
+  id: string;
+  type: 'unsupported';
+}
+
+interface UnsupportedHistogramField extends HistogramField {
+  type: Exclude<
+    KBN_FIELD_TYPES,
+    KBN_FIELD_TYPES.STRING | KBN_FIELD_TYPES.BOOLEAN | KBN_FIELD_TYPES.DATE | KBN_FIELD_TYPES.NUMBER
+  >;
+}
+
+type ChartRequestAgg = AggHistogram | AggCardinality | AggTerms;
+
+export type FieldsForHistograms = Array<
+  | NumericHistogramField
+  | NumericHistogramFieldWithColumnStats
+  | OrdinalHistogramField
+  | UnsupportedHistogramField
+>;
+
+export const fetchHistogramsForFields = async (
```

Comment:
This is exported by the package. Would be good to add some docs for it.

Reply:
Added some comments in [d3796b5](https://github.com/elastic/kibana/pull/136295/commits/d3796b51e8ceb1b88188bf1b950430b8e4797e7e).